src/metaxy/metadata_store/delta.py

```diff
@@ -0,0 +1,336 @@
+            versioning_engine_cls=PolarsVersioningEngine,
+            **kwargs,
+        )
+
+    # ===== MetadataStore abstract methods =====
+
+    def _get_default_hash_algorithm(self) -> HashAlgorithm:
+        """Use XXHASH64 by default to match other non-SQL stores."""
+        return HashAlgorithm.XXHASH64
+
+    @contextmanager
+    def _create_versioning_engine(
+        self, plan: FeaturePlan
+    ) -> Iterator[PolarsVersioningEngine]:
+        """Create Polars versioning engine for Delta store."""
+        with self._create_polars_versioning_engine(plan) as engine:
+            yield engine
+
+    @contextmanager
+    def open(self, mode: AccessMode = AccessMode.READ) -> Iterator[Self]:  # noqa: ARG002
+        """Open the Delta Lake store.
+
+        Delta-rs opens connections lazily per operation, so no connection state management needed.
+
+        Args:
+            mode: Access mode for this connection session (accepted for consistency but not used).
+
+        Yields:
+            Self: The store instance with connection open
+        """
+        # Mark store as open and validate
+        # Note: Delta auto-creates tables on first write, no need to pre-create them
+        self._is_open = True
+        self._validate_after_open()
+
+        try:
+            yield self
+        finally:
+            # delta-rs uses one-shot operations, so nothing to close
+            self._is_open = False
+
+    # ===== Internal helpers =====
+
+    def _table_name_to_feature_key(self, table_name: str) -> FeatureKey:
+        """Convert table name back to feature key.
+
+        Args:
+            table_name: Table name (directory name) to parse
+
+        Returns:
+            FeatureKey constructed from table name parts
+        """
+        # Table names are created by joining parts with "__"
+        parts = table_name.split("__")
+        # Restore system table prefix (table_name converts hyphens to underscores)
+        if parts and parts[0] == "metaxy_system":
+            parts[0] = "metaxy-system"
+        return FeatureKey(parts)
+
+    def _feature_uri(self, feature_key: FeatureKey) -> str:
+        """Return the URI/path used by deltalake for this feature."""
+        if self.layout == "nested":
+            # Nested layout: store in directories like "part1/part2/part3"
+            table_path = "/".join(feature_key.parts)
+        else:
+            # Flat layout (default): store in directories like "part1__part2__part3"
+            table_path = feature_key.table_name
+        return os.path.join(self._root_uri, table_path)
+
+    def _table_exists(self, table_uri: str) -> bool:
+        """Check whether the provided URI already contains a Delta table.
+
+        Works for both local and remote (object store) paths.
+        """
+        import deltalake  # pyright: ignore[reportMissingImports]
+        from deltalake.exceptions import (
+            TableNotFoundError,  # pyright: ignore[reportMissingImports]
```

Comment:
This may clash with our own `TableNotFoundError`, maybe import it `as DeltaTableNotFoundError`? not that it's important, just would be easier to read the code